[BEAM-9468] Fhir io

jaketf opened the pull request:
adds FhirIO healthcare connector 
Write use the executeBundle REST API method to execute transactions on the FHIR Store
Read with read / search / from notification subscription

Follow up to #1151

R: @pabloem 
CC: @lastomato, @brianlucier
------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [x] Update `CHANGES.md` with noteworthy changes.
 - [x] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | --- | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/)
Java | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/)
Python | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/) | --- | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PreCommit_Python2_PVR_Flink_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_Python2_PVR_Flink_Cron/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Python35_VR_Flink/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Python35_VR_Flink/lastCompletedBuild/) | --- | --- | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/)
XLang | --- | --- | --- | [![Build Status](https://builds.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | --- | [![Build Status](https://builds.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/)

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [![Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [![Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

jaketf commented:
Currently The R4 Integration test is failing.
Will look into this tomorrow.

jaketf commented:
Integration test data was bad due to mis-configuration of synthea exporting fhir groups.
This data is now fixed and tests are passing for all fhir versions.

jaketf commented on sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/healthcare/FhirIO.java:
```diff
@@ -0,0 +1,662 @@
+      }
+
+      @StartBundle
+      public void initBatch() throws IOException {
```
@lastomato @yeweidaniel @ygupta89 @pabloem 
Please review the logic here and LMK if you feel this is spamming the import API.
As written this will create a separate new line delimited JSON file per bundle (group of elements in a `PCollection` of arbitrary size decided by runner) and import it w/ a separate call to import API.

This was the "easiest" way to implement but may not be the most appropriate usage of the import API.

PTAL and LMK if you'd like this to refactored into something that:
1) asserts bounded PCollection
2) waits til all temp files are written 
3) blocks on import by polling the operation til complete.

It's not immediately obvious to me how to achieve 1 & 2 in the beam framework but I can do some research.

lastomato replied:
I am a little worried about the fact that the bundle size is unknown. In extreme (though unlikely) cases, we could be making one import API call per resource.

jaketf replied:
Research:

- Based on [this SO post](https://stackoverflow.com/questions/44255924/error-message-too-many-sources-provided-15285-limit-is-10000) 
We might be able to `GroupByKey` single dummy key to encourage runner to do less bundle splitting, though I'm not sure that will really do the trick. 
- From Beam docs:
> Specifically, it makes it impossible to batch any operations, such as writing elements to a sink or checkpointing progress during processing.
>The division of the collection into bundles is arbitrary and selected by the runner. This allows the runner to choose an appropriate middle-ground between persisting results after every element, and having to retry everything if there is a failure. For example, a streaming runner may prefer to process and commit small bundles, and a batch runner may prefer to process larger bundles.
([source](https://beam.apache.org/documentation/runtime/model/#bundling-and-persistence))

It seems to me that our predominant use case for Import is a batch pipeline and this sentence implies that bundles are larger in batch (not sure if we can find internal details on Dataflow Runner). 

- BigQueryIO has a [batch loads](https://github.com/apache/beam/blob/master/sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BatchLoads.java) pattern which seems similar to what we are aiming to achieve here. This seems quite complex and I'm wondering if there are simpler approaches elsewhere in the beam code base.

Do we have any guidance on how many parallel import operations would be acceptable? Perhaps we can run some benchmarks with 500, 1000, 10000 resources with this approach / adding group by single key and see if it's acceptable.

jaketf replied:
@pabloem is there a good way to signal to the runner to prefer large bundles for this DoFn? Are there any other common patterns for batching files for bulk import operations that are less complicated than the BQ File Loads (which seem to have a lot of complexity to handle dynamic destinations)?

jaketf commented on sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/healthcare/HttpHealthcareApiClient.java:
```diff
@@ -0,0 +1,572 @@
+  }
+
+  @Override
+  public HttpBody executeFhirBundle(String fhirStore, HttpBody bundle) throws IOException {
```
@lastomato @yeweidaniel @ygupta89
This is super awkward but the `executeBundle` method of client lib seems broken always giving errors about unparsable resource based on my testing (CDAP plugins also written similar to this).

Any ETA on fix for that?
Does this workaround seem acceptable til then?

lastomato replied:
There is no ETA for a fix, and it is out of our control. It is acceptable to me :)

jaketf commented:
TODOs:
- [ ] ValueProvider support
- [ ] Unit test for FhirIO dead letter handling
- [ ] Add example usage to javadoc

jaketf commented:
@lastomato 
Do you think we should add `FhirIO.Read` implementations that paginate through [fhir.search](https://cloud.google.com/healthcare/docs/reference/rest/v1beta1/projects.locations.datasets.fhirStores.fhir/search) results?

jaketf commented:
@chamikaramj FYI this is a WIP follow up to #11151 to add an IO for FHIR store.

jaketf commented:
R: @pabloem

jaketf commented:
@lastomato I added [GroupIntoBatches](https://beam.apache.org/releases/javadoc/2.19.0/org/apache/beam/sdk/transforms/GroupIntoBatches.html) in the FhirIO.Import path. 
The logic is:
- buffer `HttpBody`'s to an iterable until we have 1000 of them (this threshold was chosen arbitrarily)
- ImportFn updates the ndJson write channel with all 1000 resources
- FinishBundle will flush the batch: write to file on GCS and trigger import job

This is one way to mitigate the "import job per resource" concern but I'm open to other suggestions for achieving this.

jaketf commented:
FhirIOReadIT was flaky due to stale data (from previously run tests) sticking around in pubsub topic.
This causes the read to try to read a resource from a FHIR store that no longer exists. 
While the behavior of the IO was WAI writing these to the DLQ, I've improved the test setup to create / destroy the pubsub topic for each test.

chamikaramj commented:
Seems like one of the new tests is flaky.

https://builds.apache.org/job/beam_PostCommit_Java/5947/
https://builds.apache.org/job/beam_PostCommit_Java/5943/
https://builds.apache.org/job/beam_PostCommit_Java/5942/

chamikaramj commented:
Sorry wrong PR. Should be https://github.com/apache/beam/pull/11151

lastomato commented on sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/healthcare/FhirIO.java:
```diff
@@ -0,0 +1,977 @@
+  }
+
+  /** The type Execute bundles. */
+  public static class ExecuteBundles extends PTransform<PCollection<HttpBody>, Write.Result> {
```
Is there any specific reason that the input is HttpBody instead of String? It feels weird to me because we only use the data field in the API client.

jaketf replied:
@lastomato
My original goal was just to mimic the client library as closely as possible to be familiar to users that are used to that interface and to smooth migration to use the client library (rather than using apache http clinet / requests as a workaround). 

For example, if the signature was PTransform<PCollection<String>, Write.Result> it might be confusing to a user if they need to base 64 encode the strings themselves or not.

I was also unsure if there would be some use cases where users would want to set certain other things like extensions or content type in the HttpBody. Though it's worth noting that in the current implementation we are hard coding content type and don't handle extensions.

If you feel strongly that the API should accept Strings I'm open to refactoring for that.

jaketf replied:
discussed offline, will refactor to pass around as strings.

jaketf commented:
@pabloem @lastomato  the refactor to remove unnecessary use of `HttpBody` model (and associated coder) in favor of `String` is complete, plus an adding of HTTP status code similar to the one recently added in #11531

jaketf commented on sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/healthcare/FhirIO.java:
```diff
@@ -0,0 +1,977 @@
+        try {
+          // Blocking fhirStores.import request.
+          assert contentStructure != null;
+          Operation operation =
```
is import method idempotent?

jaketf replied:
@pabloem FYI
Yes. it is idempotent.

> The import operation is idempotent unless the input data contains multiple valid resources with the same ID but different contents. In that case, after the import completes, the store contains exactly one resource with each ID, but the duplicate entries could contain any version of the contents. For example, importing a million resources with the same ID writes only one resource to the store.
[FHIR Import Public Docs](https://cloud.google.com/healthcare/docs/concepts/fhir-import)

jaketf commented on sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/healthcare/FhirIO.java:
```diff
@@ -0,0 +1,977 @@
+        List<ResourceId> deadLetterDestinations = new ArrayList<>();
+        assert batch != null;
+        for (ResourceId file : batch) {
+          tempDestinations.add(
```
what happens if this gets rescheduled.

jaketf replied:
The steps of ImportFn are:

1. [move files](https://github.com/apache/beam/pull/11339/files/7533139cc92d4e50221b4a560937e621c9c21d8a#diff-e133729ae76dc88e72692983928b801fR765) for this batch import call to a temp prefix (in order to capture with a single wildcard for the import API call) with a [UUID](https://github.com/apache/beam/pull/11339/files/7533139cc92d4e50221b4a560937e621c9c21d8a#diff-e133729ae76dc88e72692983928b801fR740)
2. [Make and FHIR Import call](https://github.com/apache/beam/pull/11339/files/7533139cc92d4e50221b4a560937e621c9c21d8a#diff-e133729ae76dc88e72692983928b801fR770)
3. [Poll/block until this operation completes](https://github.com/apache/beam/pull/11339/files/7533139cc92d4e50221b4a560937e621c9c21d8a#diff-e133729ae76dc88e72692983928b801fR772).
4. [Clean up temp files](https://github.com/apache/beam/pull/11339/files/7533139cc92d4e50221b4a560937e621c9c21d8a#diff-e133729ae76dc88e72692983928b801fR774)

The FHIR import operation should be idempotent (see comment below).

jaketf replied:
Let's consider what happens if the process get killed / rescheduled between each of these setps
| when?                                         | GCS                                                                                                               | FHIR Store                                                                                                                                | Notes                                                                                                                                        
|-----------------------------------------------|-------------------------------------------------------------------------------------------------------------------|-------------------------------------------------------------------------------------------------------------------------------------------|-----------------------------------------------------------------------------------------------------------------------------------------------|
| (1) -> (1) between non-atomic object moves on GCS | some objects moved to a tmp-UUID prefix. | No FHIR Import Job Attempted | when rescheduled, will try to import the subset of objects that were NOT moved.
| (1) -> (2) | objects moved to a tmp-UUID prefix. | No FHIR Import Job Attempted | when rescheduled we'll fail because  the original URIs don't exist.                                                                                             
| (2) -> (3)                                    | objects moved to a tmp-UUID prefix                                                                                | an in-flight FHIR Import job that beam has lost track of. We could safely re-do this import job without consequence as it is idempotent.  | However, a rescheduled attempt will fail as the original object URIs now will not exist because they've been "claimed" by being moved to the temp dir. | 
| (3) -> (3) aka killed between polling calls:  | In some cases where InterruptedException is thrown we will move the objects in questions to the dead letter path. | in-flight FHIR Import job may or may not succeed. |                                                                                                                                               If user re-runs import from DLQ should be fine due to idempontency. But this could be wasteful if the original job in-fact succeeds |
| (3) -> (4):                                   | tmp-UUID dir will not be cleaned up properly                                                                      | FHIR Import Job succeeded  |  This is not a critical failure but inconvenient and will incur storage costs. |

jaketf replied:
We can refactor to split this up into:
1. ImportFn
    - Responsible for moving GCS files to temp dir and starting the FHIR Import
1. BlockFn
    - Responsible for blocking on each FHIR Import Operation and cleaning up GCS files
This will help our (2)-> (3), (3) -> (3) and (3) -> (4) scenarios because the OperationID and temp resource IDs will be passed between stages via `PCollection<KV<String, List<ReadableFile>>>`.

This leaves use with the (1) -> (1) and (1) -> (2) scenarios. These could be handled if we "look for" a temp dir similar to the one we'd plan on creating. Aka we look at the existing temp dirs and see if one has matching file names.

jaketf replied:
In 5540277 Refactored FhirIO.Import to:
- copy (instead of rename) files to temp dirs. 
- delete everything from gcsTempDir once window has closed.

This addresses all issues in above table except being killed between polling calls.
To mitigate this we could split into a `ImportFn` that is responsible only for starting Import jobs and  outputs KV of Operation IDs and temp dirs and `BlockTilDoneFn` which handles blocking and clean up.
@pabloem thoughts on a cleaner way to do this?

pabloem commented:
retest this please

pabloem commented on sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/healthcare/FhirIO.java:
```diff
@@ -0,0 +1,1107 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.io.gcp.healthcare;
```
Also add a line item about FhirIO to `CHANGES.md` in the root of apache/beam

pabloem commented on sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/healthcare/FhirIO.java:
```diff
@@ -0,0 +1,1107 @@
+ * href="https://cloud.google.com/healthcare/docs/concepts/fhir">Google Cloud Healthcare Fhir API.
+ * </a>
+ *
+ * <p>Read
```
```suggestion
 * <h3>Reading</h3>
```

pabloem commented on sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/healthcare/FhirIO.java:
```diff
@@ -0,0 +1,1107 @@
+ *
+ * <p>Read
+ *
+ * <p>FHIR resources can be read with {@link FhirIO.Read} supports use cases where you have a
```
```suggestion
 * <p>FHIR resources can be read with {@link FhirIO.Read}, which supports use cases where you have a
```

pabloem commented on sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/healthcare/FhirIO.java:
```diff
@@ -0,0 +1,1107 @@
+ * org.apache.beam.sdk.io.TextIO}*) .
+ *
+ * <p>Fetch Resource contents from Fhir Store based on the {@link PCollection} of message ID strings
+ * {@link FhirIO.Read.Result} where one can call {@link Read.Result#getResources()} to retrieved a
```
```suggestion
 * {@link FhirIO.Read.Result} where one can call {@link Read.Result#getResources()} to retrieve a
```

pabloem commented on sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/healthcare/FhirIO.java:
```diff
@@ -0,0 +1,1107 @@
+ * choosing. This error handling is mainly to transparently surface errors where the upstream {@link
+ * PCollection}* contains IDs that are not valid or are not reachable due to permissions issues.
+ *
+ * <p>Write Resources can be written to FHIR with two different methods: Import or Execute Bundle.
```
Perhaps add an `<h3>Writing</h3>` header

pabloem commented on sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/healthcare/FhirIO.java:
```diff
@@ -0,0 +1,1107 @@
+ *   .apply(FhirIO.readResources());
+ *
+ * // happily retrived messages
+ * PCollection<String> resources = readResult.getResources();
```
Aren't these meant to be `HealthcareIOMessage` types?

jaketf replied:
Sorry for the confusion. 

We used to pass resources around as [HttpBody](https://github.com/googleapis/google-api-java-client-services/blob/master/clients/google-api-services-healthcare/v1beta1/1.30.1/com/google/api/services/healthcare/v1beta1/model/HttpBody.java) but we only ever use the String data field so we refactored to just pass around as strings.

This eliminated the need for a HttpBodyCoder which had proven to be problematic.

If you call readResult.getFailedReads() you will get PCollection<HealthcareIOError> which is effectively a DLQ to handle if you were trying to read a resource ID that doesn't exist (or some similar error).

pabloem commented on sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/healthcare/FhirIO.java:
```diff
@@ -0,0 +1,1107 @@
+ *         .write()
+ *         .to(option.getBQFhirExecuteBundlesDeadLetterTable())
+ *         .withFormatFunction(new HealthcareIOErrorToTableRow()));
+ * }***
```
Perhaps also add code samples for doing an import?

pabloem commented on sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/healthcare/FhirIO.java:
```diff
@@ -0,0 +1,1107 @@
+   */
+  public static Import importResources(
+      String fhirStore,
+      String tempDir,
```
You may be able to make `tempDir` optional by using the `tempLocation` pipeline option internally. You can leave this for later, as it would be a backwards-compatible change to add a new factory method. But I recommend you do do this eventually, as it's nice for users to simply forget about the temporary directories.

jaketf replied:
Great suggestion! this should be easy to add!

pabloem commented on sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/healthcare/FhirIO.java:
```diff
@@ -0,0 +1,1107 @@
+       * @return the read result
+       * @throws IllegalArgumentException the illegal argument exception
+       */
+      public static FhirIO.Read.Result of(PCollectionTuple pct) throws IllegalArgumentException {
```
does this need to be public?

pabloem commented on sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/healthcare/FhirIO.java:
```diff
@@ -0,0 +1,1107 @@
+        public void processElement(ProcessContext context) {
+          String resourceId = context.element();
+          try {
+            context.output(fetchResource(this.client, resourceId));
```
this function makes RPCs to an external service one-by-one. This will be slow. It may be worth buffering these requests, and applying them with a thread pool. You can add this optimization later on, though.

jaketf replied:
The [Fhir.read](https://cloud.google.com/healthcare/docs/reference/rest/v1beta1/projects.locations.datasets.fhirStores.fhir/read) api doesn't support requesting several resources in a single call.

We could look into [Fhir.search](https://cloud.google.com/healthcare/docs/reference/rest/v1beta1/projects.locations.datasets.fhirStores.fhir/search) to request all resources with IDs in a list.

pabloem replied:
You could perform multiple `Fhir.read` requests from separate threads? This may be fine for now, though, and if you find that the extra optimization will help, then add it later on?

pabloem commented on sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/healthcare/FhirIO.java:
```diff
@@ -0,0 +1,1107 @@
+      }
+
+      /** DoFn for fetching messages from the Fhir store with error handling. */
+      public static class StringGetFn extends DoFn<String, String> {
```
This may not need to be public?

pabloem commented on sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/healthcare/FhirIO.java:
```diff
@@ -0,0 +1,1107 @@
+     *       error message and stacktrace.
+     * </ul>
+     */
+    public static class FetchResourceJsonString
```
May not need to be public?

pabloem commented on sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/healthcare/FhirIO.java:
```diff
@@ -0,0 +1,1107 @@
+
+    public static Write fhirStoresImport(
+        ValueProvider<String> fhirStore,
+        ValueProvider<String> gcsTempPath,
```
as mentioned before, you may be able to have a constructor without gcsTempPath

pabloem commented on sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/healthcare/FhirIO.java:
```diff
@@ -0,0 +1,1107 @@
+      PCollection<HealthcareIOError<String>> failedImports;
+      switch (this.getWriteMethod()) {
+        case IMPORT:
+          LOG.info(
```
A warning may work here. Up to you, though. sorry about the back-and-forth.  : )

pabloem commented on sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/healthcare/FhirIO.java:
```diff
@@ -0,0 +1,1107 @@
+    }
+
+    @Override
+    public Write.Result expand(PCollection<String> input) {
```
TODO(@pabloem) take a look at this function

pabloem commented on sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/healthcare/FhirIO.java:
```diff
@@ -0,0 +1,1107 @@
+        try {
+          // Validate that data was set to valid JSON.
+          mapper.readTree(body);
+          client.executeFhirBundle(fhirStore, body);
```
This is executing a single RPC per element, right? This will likely be slow, and lead to underutilization of worker resources. As a follow up, messages could be buffered, and RPCs issued via a thread pool

jaketf replied:
IMO Batching should be handled upstream from the connector based on what type of bundles the user is trying to commit to fire.

executeBundle is a single call per (beam) element (but each (beam) element in this case is already a [FHIR Bundle](https://www.hl7.org/fhir/bundle.html)) but in a transaction type bundle it's very crucial what elements are batched together (as they are committed as a single transaction on the back-end). The user must control this to get the desired behavior.

with batch type bundle the operations are executed individually on the back end (but provided by the client as a single call). 
We could provide a convenience PTransform for buffering batches into a bundle. 

context of what we support in Google HCAPI:
https://cloud.google.com/healthcare/docs/reference/rest/v1beta1/projects.locations.datasets.fhirStores.fhir/executeBundle

broader context of what the FHIR spec has for bundle type:
https://www.hl7.org/fhir/valueset-bundle-type.html

jaketf replied:
To be a bit clearer:

- FHIR bundle is 1:1 with beam element here (but may contain many operations). 
- The FHIR bundle a field that dictates if it is transaction or batch type but they use the same endpoint / method.

pabloem replied:
Makes sense. Sounds good to me!

pabloem commented on sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/healthcare/HL7v2MessageCoder.java:
```diff
@@ -29,7 +29,15 @@
+    return new HL7v2MessageCoder();
+  }
+
+  public static HL7v2MessageCoder of(Class<HL7v2Message> clazz) {
```
Howcome you receive a class but ignore it?

jaketf replied:
My understanding was this is necessary in order to use `@Default.Coder(HL7v2MessageCoder)` decorator on the `HL7v2Message` class.

However, I forgot to add that decorator so no wonder this looks confusing, good catch!
 
https://beam.apache.org/documentation/programming-guide/#default-coders-and-the-coderregistry

pabloem replied:
Ah you're right  : ) - cool.

pabloem commented:
retest this please

pabloem commented on sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/healthcare/FhirIO.java:
```diff
@@ -0,0 +1,1144 @@
+              .get(Write.TEMP_FILES)
+              .apply(
+                  "Shard files", // to paralelize group into batches
+                  WithKeys.of(ThreadLocalRandom.current().nextInt(0, numShards)))
```
This is providing a static key, not a key-generating function, right?
```suggestion
                  WithKeys.of(elm -> ThreadLocalRandom.current().nextInt(0, numShards)))
```
(I think)

pabloem commented:
retest this pleas

pabloem commented:
retest this please

pabloem commented:
retest this please

pabloem commented:
Run Java PostCommit

pabloem commented:
Run Java PostCommit

pabloem commented:
I'm unsure if the failures are due to an issue with this PR. Feel free to inspect:
https://scans.gradle.com/s/c4wwwmeidugrc/failure#top=0

pabloem commented:
Run Java PreCommit

pabloem commented:
Fhir postcomits are failing: https://builds.apache.org/job/beam_PostCommit_Java_PR/362/

jaketf commented:
> I'm unsure if the failures are due to an issue with this PR. Feel free to inspect:
> https://scans.gradle.com/s/c4wwwmeidugrc/failure#top=0

There's not much information in that scan. Running
`./gradlew checkstyleMain checkstyleTest` locally there was an issue in FhirIO
```
> Task :sdks:java:io:google-cloud-platform:checkstyleMain FAILED
[ant:checkstyle] [ERROR] /usr/local/google/home/jferriero/VersionControl/beam/sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/healthcare/FhirIO.java:846:42: Name 'LOG' must match pattern '^[a-z][a-zA-Z0-9]*$'. [MemberName]
```
I'll clean this up. 
However there were many failures due to FlinkState internals

> Fhir postcomits are failing: https://builds.apache.org/job/beam_PostCommit_Java_PR/362/

ACK looking into this.

pabloem commented:
I see at least this one (
https://builds.apache.org/job/beam_PostCommit_Java_PR/362/testReport/junit/org.apache.beam.sdk.io.gcp.healthcare/FhirIOReadIT/testFhirIORead_DSTU2_/)
with a Dataflow job error. Perhaps once that's fixed, other issues will be
fixed.

On Thu, May 14, 2020 at 12:10 PM Jacob Ferriero <notifications@github.com>
wrote:

> I'm unsure if the failures are due to an issue with this PR. Feel free to
> inspect:
> https://scans.gradle.com/s/c4wwwmeidugrc/failure#top=0
>
> There's not much information in that scan. Running
> ./gradlew checkstyleMain checkstyleTest locally there was an issue in
> FhirIO
>
> > Task :sdks:java:io:google-cloud-platform:checkstyleMain FAILED
> [ant:checkstyle] [ERROR] /usr/local/google/home/jferriero/VersionControl/beam/sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/healthcare/FhirIO.java:846:42: Name 'LOG' must match pattern '^[a-z][a-zA-Z0-9]*$'. [MemberName]
>
> I'll clean this up.
> However there were many failures due to FlinkState internals
>
> Fhir postcomits are failing:
> https://builds.apache.org/job/beam_PostCommit_Java_PR/362/
>
> ACK looking into this.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/beam/pull/11339#issuecomment-628833485>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAJ5Z3FITPXYPB3WICDP2HLRRQ62HANCNFSM4MDSGDLQ>
> .
>

jaketf commented:
Yeah they are  parameterized tests that are all failing like this:
```
WARNING: No terminal state was returned within allotted timeout. State value RUNNING
May 14, 2020 2:37:23 AM org.apache.beam.runners.dataflow.TestDataflowRunner waitForStreamingJobTermination
INFO: Dataflow job 2020-05-13_19_27_22-9842849986096969383 took longer than 600 seconds to complete, cancelling.
May 14, 2020 2:37:23 AM org.apache.beam.runners.dataflow.TestDataflowRunner run
WARNING: Dataflow job 2020-05-13_19_27_22-9842849986096969383 did not output a success or failure metric.
May 14, 2020 2:37:24 AM io.grpc.internal.ManagedChannelOrphanWrapper$ManagedChannelReference cleanQueue
SEVERE: *~*~*~ Channel ManagedChannelImpl{logId=15, target=pubsub.googleapis.com:443} was not shutdown properly!!! ~*~*~*
    Make sure to call shutdown()/shutdownNow() and wait until awaitTermination() returns true.
java.lang.RuntimeException: ManagedChannel allocation site
```

jaketf commented:
[FhirIOReadIT](https://github.com/apache/beam/pull/11339/files#diff-7a1359c60a094e73275769adb69b35d3R57) borrows the [TestPubsubSignal](https://github.com/apache/beam/blob/44f3d3f2d52e93a34b05068bf76f6f9d2611bf77/sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/pubsub/TestPubsubSignal.java) pattern from [PubsubReadIT](https://github.com/apache/beam/blob/a9e14ff7f4b1aafa9915cb95e1bb7e7d3ab6a28b/sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/pubsub/PubsubReadIT.java#L41)
This works fine on direct runner for me. 
It seems this is a dataflow runner / test pubsub signal thing.
I will look into if I'm not closing some channel in my tests.
 
@pabloem Have we seen anything flaky on this before?

jaketf commented:
There is a separate crop of issues of 

```java.lang.NoClassDefFoundError: Could not initialize class org.apache.beam.sdk.io.gcp.healthcare.FhirIOTestUtil```

In the execute bundles test. Are test utility classes not bundled up and set to dataflow? should I move this under main in the source tree?

pabloem commented:
afaik, the test classes should be bundled together (see https://github.com/apache/beam/blob/4c7d5460ba1d643e7fd57fa8f2f8e6a87d80bee5/sdks/java/io/google-cloud-platform/build.gradle#L118-L122)

@mwalenia do you know why the testutil classes could be missing from the IT test?

jaketf commented:
- [x] Fixed javadoc checkstyle issue
- [x] Added pubsub.close() to the @After in FhirIOReadIT (this should fix the pubsub not shutdown properly message)
- [x] Added KV coders to FhirIO.Import

see pre-commit output here.
https://gist.github.com/jaketf/e7c9116daed60e7babffdb4c99ae1d54

jaketf commented:
@pabloem  I'm not sure what's going on with the class loading IT issue but I think we could re-run the precommit

pabloem commented:
retest this please

mwalenia commented:
@pabloem everything should be bundled, `build.gradle` for `io.google-cloud-platform` contains everything that's needed. Maybe there's some other error?

@jaketf Can you elaborate on the `java.lang.NoClassDefFoundError: Could not initialize class org.apache.beam.sdk.io.gcp.healthcare.FhirIOTestUtil`? What exactly causes this error?

>Are test utility classes not bundled up and set to dataflow?

what do you mean by this?

pabloem commented:
We have a failure from a test checking out the API surface for the gcp io package: https://builds.apache.org/job/beam_PreCommit_Java_Commit/11386/testReport/junit/org.apache.beam.sdk.io.gcp/GcpApiSurfaceTest/testGcpApiSurface/

pabloem commented:
this test checks that we're not increasing the dependencies that we expose in the API. We can add the Healthcare IO classes.
Other classes I think we should try to figure out how to avoid exposing them. It seems that they mostly come from the `HttpHealthcareApiClient$HealthcareHttpException`

jaketf commented:
@mwalenia 
> What exactly causes this error?https://builds.apache.org/job/beam_PostCommit_Java_PR/362/testReport/junit/org.apache.beam.sdk.io.gcp.healthcare/FhirIOWriteIT/testFhirIO_ExecuteBundle_R4_/

potentially there was an issue staging all appropriate files
```
May 14, 2020 2:05:45 AM org.apache.beam.runners.dataflow.options.DataflowPipelineOptions$StagingLocationFactory create
INFO: No stagingLocation provided, falling back to gcpTempLocation
May 14, 2020 2:05:45 AM org.apache.beam.runners.dataflow.DataflowRunner fromOptions
INFO: PipelineOptions.filesToStage was not specified. Defaulting to files from the classpath: will stage 183 files. Enable logging at DEBUG level to see which files will be staged.
May 14, 2020 2:05:46 AM org.apache.beam.runners.dataflow.options.DataflowPipelineOptions$StagingLocationFactory create
INFO: No stagingLocation provided, falling back to gcpTempLocation
May 14, 2020 2:05:46 AM org.apache.beam.runners.dataflow.DataflowRunner fromOptions
INFO: PipelineOptions.filesToStage was not specified. Defaulting to files from the classpath: will stage 183 files. Enable logging at DEBUG level to see which files will be staged.
May 14, 2020 2:05:46 AM org.apache.beam.runners.dataflow.options.DataflowPipelineOptions$StagingLocationFactory create
INFO: No stagingLocation provided, falling back to gcpTempLocation
May 14, 2020 2:05:46 AM org.apache.beam.runners.dataflow.DataflowRunner fromOptions
INFO: PipelineOptions.filesToStage was not specified. Defaulting to files from the classpath: will stage 183 files. Enable logging at DEBUG level to see which files will be staged.
```

jaketf commented:
@pabloem re: api surface.
So the history here is that the client library executeBundle method is broken so we were forced to use out own http client (and our own exceptions).

I think I can make the constructor / factory that include apache http deps private as they wouldn't be constructed by user. I'll take a look.

jaketf commented:
@pabloem 843104d was a pretty simple fix for api surface test
```
git rev-parse HEAD && ./gradlew :sdks:java:io:google-cloud-platform:cleanTest :sdks:java:io:google-cloud-platform:test --tests "org.apache.beam.sdk.io.gcp.GcpApiSurfaceTest"
843104dacaa9a7064247d6870032c9d69d51afe8
Configuration on demand is an incubating feature.

> Task :sdks:java:io:google-cloud-platform:compileTestJava
Note: Some input files use or override a deprecated API.
Note: Recompile with -Xlint:deprecation for details.
Note: Some input files use unchecked or unsafe operations.
Note: Recompile with -Xlint:unchecked for details.

Deprecated Gradle features were used in this build, making it incompatible with Gradle 6.0.
Use '--warning-mode all' to show the individual deprecation warnings.
See https://docs.gradle.org/5.2.1/userguide/command_line_interface.html#sec:command_line_warnings

BUILD SUCCESSFUL in 13s
68 actionable tasks: 2 executed, 66 up-to-date
```

pabloem commented:
retest this please

pabloem commented:
retest this please

pabloem commented:
retest this please

pabloem commented:
retest this please

pabloem commented:
Run Java PostCommit

pabloem commented:
Run Java PostCommit

pabloem commented:
Run Java PreCommit

pabloem commented:
Run Java PreCommit

pabloem commented:
Run Java PreCommit

pabloem commented:
Run Java PreCommit

pabloem commented:
Run Java PreCommit

pabloem commented:
retest this please

pabloem commented:
retest this please

pabloem commented:
Run Java PostCommit

pabloem commented:
Run Java PostCommit

pabloem commented:
Run Java PostCommit

pabloem commented:
retest this please

pabloem commented:
Run Java PostCommit

pabloem commented:
retest this please

pabloem commented:
retest this please

pabloem commented:
Run Java PostCommit

pabloem commented:
retest this please

pabloem commented:
retest this please

pabloem commented:
Run Java_Examples_Dataflow PreCommit

pabloem commented:
Run Java_Examples_Dataflow PreCommit

pabloem commented:
Run Java PreCommit

jaketf commented:
The issue with FhirIOReadIT seems to be some misuse of [TestPubsubSignal](https://beam.apache.org/releases/javadoc/2.20.0/org/apache/beam/sdk/io/gcp/pubsub/TestPubsubSignal.html)
[Example Job](https://pantheon.corp.google.com/dataflow/jobs/us-central1/2020-05-19_19_05_19-1057064265214622054?project=apache-beam-testing) clearly has elements added to the "waitForAnyMessage" task
but the success signal never gets published to the results topic.

This is tangential to the behavior being tested. Investigating other means of performing this test.

jaketf commented:
I think that something about `TestPubsubSignal` does not play well with the way this integration test suite runs on Dataflow.

Following in suit with `PubsubReadIT` (which also uses `TestPububSignal`) `BigQueryIOReadIT` and `BigQueryIOStorageReadTableRowIT` I've added this to the exclude of this project's build file.

This test can still be run (and passes) using DirectRunner and the evidence in the [above comment](https://github.com/apache/beam/pull/11339#issuecomment-631203002) points to this being a red herring of an issue with test pubsub signal and dataflow runner. 

Potentially related issue: [BEAM-6804](https://issues.apache.org/jira/browse/BEAM-6804?jql=text%20~%20%22TestPubsubSignal%22)

I've filed an issue for this instance.

@pabloem please let me know if this is acceptable.

pabloem commented:
Retest this please

pabloem commented:
Run Java PostCommit

pabloem commented:
Run Java PreCommit

pabloem commented:
Run Java PostCommit

pabloem commented:
Run Java PostCommit

pabloem commented:
Run Java PostCommit

pabloem commented:
retest this please

pabloem commented:
retest this please

pabloem commented:
Run Java Postcommit

TheNeuralBit commented:
This PR is almost 300k lines, it seems to mostly be due to committing a lot of JSON test data, which isn't ideal. Is there any way to store that test data elsewhere?

jaketf commented:
We could stage these / pull these in from GCS.
This will mean devs would need access to this bucket to run the tests.

TheNeuralBit commented:
@pabloem I'm fine with adding the test data in this PR if you are. Sorry to jump in with a random comment.

pabloem commented:
heh no worries. I'll discuss with @jaketf to see what we do...

pabloem commented:
retest this please

pabloem commented:
retest this please

pabloem commented:
retest this please

pabloem commented:
retest this please

pabloem commented:
after looking around, it seems like it's a fairly common pattern to have test files committed to source control as resources. These aren't huge, so I'll merge them.

pabloem commented:
retest this please

TheNeuralBit commented:
retest this please

pabloem commented:
retest this please

TheNeuralBit commented:
> after looking around, it seems like it's a fairly common pattern to have test files committed to source control as resources. These aren't huge, so I'll merge them.

Ok, as long as there's precedent for it. And you're right these files aren't that big as far as test data goes, it just looks a little scary to see a 300k line diff since they're human readable

pabloem commented:
I found these examples:
https://softwareengineering.stackexchange.com/questions/212639/should-i-store-test-files-in-source-control
https://softwareengineering.stackexchange.com/questions/257881/should-test-data-be-checked-into-version-control
https://stackoverflow.com/questions/3060764/tfs-testing-and-placement-of-test-files-where
https://stackoverflow.com/questions/28253556/what-is-a-right-way-to-deal-with-tests-temporary-files-and-version-control

pabloem commented:
retest this please

pabloem commented:
retest this please

pabloem commented:
retest this please

jaketf commented:
Seems pre-commit is stuck reporting back to github.
There's one unrelated test that failed
https://builds.apache.org/job/beam_PreCommit_Java_Commit/11508/testReport/org.apache.beam.sdk.extensions.ml/VideoIntelligenceIT/annotateVideoFromURINoContext/

TheNeuralBit commented:
That definitely seems like a flake to me

jaketf commented:
are we able to run post commit?

pabloem commented:
Run Java PostCommit

pabloem commented:
Run Java PreCommit

pabloem commented:
Run Java PreCommit

jaketf commented:
🚀

y1chi commented:
@jaketf the integration tests seem to be flaky and breaks postcommit sometimes, I opened https://issues.apache.org/jira/browse/BEAM-10419 do you mind take a look?
